src/vs/vscode.proposed.d.ts

```diff
@@ -1524,15 +1525,17 @@ declare module 'vscode' {
 		data: { [key: string]: any };
 	}
 
-	export type CellOutput = CellStreamOutput | CellErrorOutput | CellDisplayOutput;
+	export type CellOutput = CellStreamOutput | CellErrorOutput | CellDisplayOutput & { metadata: { [key: string]: any} }; // Metadata can be put here by an execution so that a renderer can use it
 
 	export interface NotebookCell {
 		readonly uri: Uri;
 		handle: number;
 		language: string;
 		cellKind: CellKind;
 		outputs: CellOutput[];
-		getContent(): string;
+		metadata: { [key: string]: any }
```

Comment:
Looking at this structure, the `NotebookCell` is mutable.

* Should we instead have a separate way of adding metadata to cells?
* Cannot the extension authors add metadata in their own extension and have it keyed to the notebookCell? This way VSC is free of storing metadata?
* If the metadata is unstructured, its not really useful to other extension authors, hence see little benefit here

Reply:
Some metadata is valid, such as:
* Readonly
* Selected

Though, they could be upgraded to standard properties on the cell..So other extensions can utilize them...

Reply:
Yeah but then one extension cannot communicate metadata to another cell. Or in the case we have now, one product to another.

Reply:
> we have now, one product to another.

Hmm, doesn't feel right. Its basically `any`, hence to me `no defined api`.
But do we have a use case of where we'd like one extension to expose metadata to another.

I can only think of gather, is that the usecase.

Reply:
User puts metadata in the jupyter notebook editor. We have a custom renderer that reads it and uses it for output. 

It's also where the kernel spec information would go so we'd know what kernel to start. Otherwise we're predefining what a kernel is and that's a jupyter specific thing.

Reply:
Not sure we need the concept of execution.

Why not?:
* Extensions provide results and tells VSC what to render
* VSC then fires events saying `onResult` or `onRendered` or similar
    * Now other extensions can listen in on them and know something has happened...
    * Similarly VSC can fire onEdited events...
* Extensions provide results for document wide results as `Document Content`. VSC Doesn't need to know whether its tied to a cell or not, all VSC needs to know is what needs to be rendered, just as with Cell results. 

It just deals with rendering something. Extensions execute ells and provide output. The missing bit is hooks for buttons, those handlers are what cause extension authors to provide results..

Reply:
It has to be any so it can not be bound to the jupyter specific implementation.

Reply:
> It's also where the kernel spec information would go so we'd know what kernel to start.

VSC doesn't need this. Extension authors can get this from the file or notebook, we do'nt need VSC API to provide this. I.e. only extensions need this information.

Reply:
Execution providers don't have access to the notebook (unless they read it from disk?) And how would a change in the kernel be propagated around?

Reply:
The notebookProvider translates the json into the NotebookDocument format. An execution provider can then execute the notebook. Hmm, maybe metadata doesn't work here either, but rather a more concrete ExecutionInfo or something along those lines.

However the other metadata in a notebook or cell that updates in memory, how would that be communicated between extensions?

Reply:
> And how would a change in the kernel be propagated around?

Why do we need to propagate this to other extensions?
I'm assuming VSC will provide access to the notebook data (raw string) or similar (or Uri).

Reply:
> how would that be communicated between extensions?

Not sure why extensions need them? I'd like to understand a usecase.
As suggested earlier, why can't extensions provide metadata along with results when calling VSC to render results.
Then VSC can propagate this to other extensions saying `here you go I received these results+metadata, do what you want with it`

Reply:
Extension 1: NotebookProvider = turns .foobar files into Notebooks
Extension 2: Renders results and looks at the 'collapsed/expanded' metadata that might be on a cell.
Extension 3: Executes notebooks. Knows nothing of metadata, just passes it along.

Reply:
I agree I don't like the metadata: any thing either. Trying to think of some other way to communicate information in the Notebook (both in memory and from disk) to other extensions

Reply:
Kernel spec information might be specific enough we create a VS code euphemism for it. I think I'll try that. But other metadata is weirder.

Reply:
> 'collapsed/expanded' 

These are some of the generic items that will have to be exposed by VSC. I.e. it is UI state, hence VSC will have to expose them to extension authors vs extension authors putting some randomnly named metadata to indicate a cell is expanded or collapsed, After all that initial information needs to come from VSC.
I.e. `collpased/expanded/selected/focused/readonly` and others would be VSC specific and I'd expect to have specific properties on cells for these.

Reply:
> communicate information in the Notebook (both in memory and from disk) to other extensions

Still feel results would provide enough information that can be shared with other extensions.

Reply:
You mean the metadata tag I added to every output?

Reply:
That might be enough. It means a NotebookProvider can't communicate any data in the base notebook though.

Reply:
> You mean the metadata tag I added to every output?

That's great.
I think we need a way for VSC to broadcast this information to other extensions.

Reply:
They would all get it. NotebookExecutionObserver would get it on its different callbacks. The NotebookOutputTransformer would get it when transforming and the NotebookOutputRenderer would see it when rendering. 

I'm still worried about all the extra information at the bottom of an ipynb. What if you need that to start your kernel? Or your language server? I guess you could read it from disk too though.

Reply:
> I guess you could read it from disk too though.

I would except VSC to provide raw access to the file or provide a Uri. At the end of the day `ipynb` is one format that backs notbooks. But it doesn't have to be an `ipynb` file.

Reply:
I'm not saying we throw away ipynb, just saying that VSC will provide some way to access the information, at a minimum the Uri (even if its remote we should be able to get the data when running remotely).